odxtools/preconditionstateref.py

```diff
@@ -0,0 +1,142 @@
+            in_param_if_snref=in_param_if_snref,
+            in_param_if_snpathref=in_param_if_snpathref,
+            **kwargs)
+
+    def __post_init__(self) -> None:
+        if self.value is not None:
+            odxassert(self.in_param_if_snref is not None or self.in_param_if_snref is not None,
+                      "If VALUE is specified, a parameter must be referenced")
+
+    def _build_odxlinks(self) -> Dict[OdxLinkId, Any]:
+        return {}
+
+    def _resolve_odxlinks(self, odxlinks: OdxLinkDatabase) -> None:
+        self._state = odxlinks.resolve(self, State)
+
+    def _resolve_snrefs(self, context: SnRefContext) -> None:
+        pass
+
+    def applies(self, state_machine: "StateMachine", params: List[Parameter],
+                param_value_dict: ParameterValueDict) -> bool:
+        """Given a state machine, evaluate whether the precondition is fulfilled or not
+
+        Note that the specification is unclear about what the
+        parameters are: It says "The optional VALUE together with the
+        also optional IN-PARAM-IF snref at STATE-TRANSITION-REF and
+        PRE-CONDITION-STATE-REF can be used if the STATE-TRANSITIONs
+        and pre-condition STATEs are dependent on the values of the
+        referenced PARAMs.", but it does not specify what the
+        "referenced PARAMs" are. For the state transition refs, let's
+        assume that they are the parameters of a positive response
+        received from the ECU, whilst we assume that they are the
+        parameters of the request for PRE-CONDITION-STATE-REFs.
+        """
+
+        if state_machine.active_state != self.state:
+            # if the active state of the state machine is not the
+            # specified one, the precondition does not apply
+            return False
+
+        if self.in_param_if_snref is None and self.in_param_if_snpathref is None:
+            # no parameter given -> only the specified state is relevant
+            return True
+
+        param, param_value = \
+            _resolve_in_param(self.in_param_if_snref, self.in_param_if_snpathref, params, param_value_dict)
+
+        if param is None:
+            # The referenced parameter does not exist. TODO: is this an error?
+            return False
+        elif not isinstance(
+                param,
+            (CodedConstParameter, PhysicalConstantParameter, TableKeyParameter, ValueParameter)):
+            # see checker rule 194 in section B.2 of the spec
+            odxraise(f"Parameter referenced by state transition ref is of "
+                     f"invalid type {type(param).__name__}")
+            return False
+        elif isinstance(param, (CodedConstParameter, PhysicalConstantParameter,
+                                TableKeyParameter)) and self.value is not None:
+            # see checker rule 193 in section B.2 of the spec. Why can
+            # no values for constant parameters be specified? (This
+            # seems to be rather inconvenient...)
```

Comment:
Because that would mean that the state will always apply, or always won't apply, it's a logic constraint, no need to hardcode it in the code

Reply:
yes, but for state transitions multiple responses could apply, so if they use a constant exhibiting the same name for e.g. their error code, they cannot be distinguished...